### PR TITLE
Paper handler diagnostic

### PR DESCRIPTION
**Pull request by kshen0**
Recommend reviewing by commit.

## Overview
https://github.com/votingworks/vxsuite/issues/4718

Adds a diagnostic flow for the paper handler.

## Demo Video or Screenshot

https://github.com/votingworks/vxsuite/assets/1060688/33ab6d57-5f19-4c53-af9d-533527573555

Screenshot 2024-04-22 at 6 03 32 PM

Screenshot 2024-04-22 at 6 03 38 PM

## Testing Plan
- added tests

## Checklist

- [x] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [x] I have added a screenshot and/or video to this PR to demo the change
- [x] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: apps/mark-scan/backend/src/custom-paper-handler/test_utils.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { Buffer } from 'buffer';
 const BALLOT_FIXTURE_FRONT_PATH = join(
   __dirname,
   'fixtures',
-  'ballot-pdf-data.bin'
```
Comment (kshen0) [inline apps/mark-scan/backend/src/custom-paper-handler/test_utils.ts:12]: Same ballot with a large top margin to more realistically represent markscan ballots. `ballot-pdf-data.bin` is now only used by the state machine CLI which is unmaintained and should be removed in a separate PR.

File: libs/fixtures/src/index.ts [hunk 2]
```diff
@@ -11,6 +11,10 @@ export {
   electionDefinition as electionGeneralDefinition,
   election as electionGeneral,
 } from './data/electionGeneral/election.json';
+export {
+  electionDefinition as electionMarkScanDiagnosticDefinition,
+  election as electionMarkScanDiagnostic,
+} from './data/electionGeneral/election.json';
```
Comment (kshen0) [inline libs/fixtures/src/index.ts:17]: This is a clone of `electionGeneralDefinition` so that fixture can be changed without impact to the diagnostic

File: libs/ui/src/ui_strings/app_strings.tsx [hunk 3]
```diff
@@ -359,7 +359,7 @@ export const appStrings = {
   instructionsBmdLoadPaper: () => (
     <UiString uiStringKey="instructionsBmdLoadPaper">
       Please feed one sheet of paper into the front input tray. The printer will
-      automatically grab the paper when positioned correctly.
```
Comment (kshen0) [inline libs/ui/src/ui_strings/app_strings.tsx:362]: Somehow this felt a little aggressive

File: apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx [hunk 4]
```diff
@@ -112,10 +112,16 @@ export function DiagnosticsScreen({
...
+              paperHandlerSectionChildren={
+                <Button
+                  onPress={() => {
+                    startPaperHandlerDiagnosticMutation.mutate();
+                    history.push('/paper-handler');
+                  }}
+                >
+                  Test Printer/Scanner
+                </Button>
+              }
```
Comment (kshen0) [inline apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx:124]: Per @adghayes this is the preferred pattern for excluding the `Test` buttons from the readiness report (ie. we just omit the `Children` props in the readiness report)

File: libs/fixtures/src/data/electionFamousNames2021/bmd-ballot-complete-p1.jpg.ts
Comment (kshen0) [inline libs/fixtures/src/data/electionFamousNames2021/bmd-ballot-complete-p1.jpg.ts:1]: This change was included with running `pnpm build:resources`

Comment (kshen0) [inline libs/fixtures/src/data/electionFamousNames2021/bmd-ballot-complete-p1.jpg.ts:1]: This change was a result of running `pnpm build:resources` in `libs/fixtures`

File: libs/fixtures/src/index.ts [hunk 5]
```diff
@@ -11,10 +11,6 @@ export {
   electionDefinition as electionGeneralDefinition,
   election as electionGeneral,
 } from './data/electionGeneral/election.json';
-export {
-  electionDefinition as electionMarkScanDiagnosticDefinition,
-  election as electionMarkScanDiagnostic,
-} from './data/electionGeneral/election.json';
```
Comment (kshen0) [inline libs/fixtures/src/index.ts]: This should have been `./data/electionMarkScanDiagnostic/...` but this fixes the issue by just cloning `electionGeneral` to `mark-scan/backend`

File: apps/mark-scan/backend/package.json [hunk 6]
```diff
@@ -8,7 +8,7 @@
     "build"
   ],
   "scripts": {
-    "build": "tsc --build tsconfig.build.json && copyfiles -u 3 src/custom-paper-handler/fixtures/*.jpg build/custom-paper-handler/fixtures",
+    "build": "tsc --build tsconfig.build.json && copyfiles -u 3 src/custom-paper-handler/ballot-assets/*.* build/custom-paper-handler/ballot-assets",
```
Comment (kshen0) [inline apps/mark-scan/backend/package.json:11]: These ballot assets are required in prod and need a separate change to prod image build to ensure they make it there. We explicitly want to differentiate `fixtures` for dev from `ballot assets` for prod.

File: libs/ui/src/diagnostics/mark_scan_paper_handler_section.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+export function MarkScanPaperHandlerSection({
+  mostRecentPaperHandlerDiagnostic,
+  isPaperHandlerDetected,
+  onStartPaperHandlerDiagnostic,
+}: MarkScanPaperHandlerSectionProps): JSX.Element {
+  if (mostRecentPaperHandlerDiagnostic) {
+    assert(mostRecentPaperHandlerDiagnostic.type === 'mark-scan-paper-handler');
+  }
+
+  return (
```
Comment (kshen0) [inline libs/ui/src/diagnostics/mark_scan_paper_handler_section.tsx:22]: I'll probably create a common component in the PR that adds the PAT diagnostic based on what similarities emerge between all the diagnostics.

File: libs/ui/src/main.tsx [hunk 8]
```diff
@@ -1,12 +1,14 @@
...
 import { assert } from '@votingworks/basics';
 import { SizeMode } from '@votingworks/types';
+import { JustifyContent } from './layout';
 
 export interface MainProps {
   padded?: boolean;
   centerChild?: boolean;
   flexRow?: boolean;
   flexColumn?: boolean;
+  justifyContent?: JustifyContent;
```
Comment (kshen0) [inline libs/ui/src/main.tsx:11]: Used here https://github.com/votingworks/vxsuite/pull/4752/commits/5e3f79b14c78a62fe48d9efefe29bfab6291a2df#diff-362a9f3be9cd78b838401516937795ef7351c7037c6671e56a861f59a121f961R82

File: apps/mark-scan/frontend/src/pages/pat_device_identification/pat_device_identification_page.tsx [hunk 9]
```diff
@@ -94,7 +94,7 @@ export function PatDeviceIdentificationPage({
             {statusStrings[currentStepId]}
           </P>
         </DiagnosticScreenHeader>
-        <StepContainer>{steps[currentStepId]}</StepContainer>
+        <PatStepContainer>{steps[currentStepId]}</PatStepContainer>
```
Comment (kshen0) [inline apps/mark-scan/frontend/src/pages/pat_device_identification/pat_device_identification_page.tsx:97]: The PAT "diagnostic" here is a calibration flow for the voter. It still needs to be wired up to the `diagnostic_record` construct for system admins.

File: apps/mark-scan/frontend/src/api.ts [hunk 10]
```diff
@@ -167,14 +167,33 @@ export const getApplicationDiskSpaceSummary = {
   },
 } as const;
 
+export const getMostRecentPaperHandlerDiagnostic = {
+  queryKey(): QueryKey {
+    return ['getMostRecentPaperHandlerDiagnostic'];
```
Comment (kshen0) [inline apps/mark-scan/frontend/src/api.ts:172]: In a future PR `getMostRecentAccessibleControllerDiagnostic` and `getMostRecentPaperHandlerDiagnostic ` should probably be combined into `getMostRecentDiagnostic(input: { diagnosticType: DiagnosticType })`  and queries invalidated by query param.

File: .circleci/config.yml [hunk 11]
```diff
@@ -43,6 +43,8 @@ jobs:
             JEST_JUNIT_OUTPUT_DIR: ./reports/
       - store_test_results:
           path: apps/admin/backend/reports/
+      - store_artifacts:
+          path: apps/admin/backend/src/__image_snapshots__/__diff_output__/
```
Comment (kshen0) [inline .circleci/config.yml:47]: Snapshot test diffs were previously not available in circleci

File: apps/mark-scan/frontend/src/apimachine_config.test.tsx [hunk 12]
```diff
@@ -27,7 +27,7 @@ test('machineConfig is fetched from api client by default', async () => {
   apiMock.expectGetElectionState({
     precinctSelection: ALL_PRECINCTS_SELECTION,
   });
-  render(<App reload={jest.fn()} apiClient={apiMock.mockApiClient} />);
```
Comment (kshen0) [inline apps/mark-scan/frontend/src/apimachine_config.test.tsx:30]: This change and others below it are a result of merging main (specifically https://github.com/votingworks/vxsuite/pull/4781) into this branch and then rebasing. Sorry for the clutter :/

File: apps/mark-scan/frontend/src/app_root.tsx
Comment (kshen0) [inline apps/mark-scan/frontend/src/app_root.tsx:1]: Can ignore this commit - result of merging main and rebasing

File: apps/mark-scan/backend/src/custom-paper-handler/electionPaperHandlerDiagnostic/election.json.ts
Comment (jonahkagan) [inline apps/mark-scan/backend/src/custom-paper-handler/electionPaperHandlerDiagnostic/election.json.ts:1]: How will this file get updated if we need to change its content? Or if the election definition format changes? I think it will be easier to update if we use a raw JSON file rather than a compiled TS file.

It also feels a little weird to use the fixture candidate and contest names in this diagnostic ballot. I would feel better about something that's more obviously for testing (e.g. in the VxAdmin printer diagnostic we use "Mock Precinct" as the precinct name in the [report we print](https://github.com/votingworks/vxsuite/blob/81a6b48bff23abb68ac3cdc5512544e0589e42c1/apps/admin/backend/src/__image_snapshots__/test-print-1.png))

File: apps/mark-scan/backend/package.json (hunk 6 above)
Comment (jonahkagan) [reply apps/mark-scan/backend/package.json:11]: I'm worried that it will be difficult to keep static ballot assets up to date as we make changes to code. It's a problem that's been a real annoyance in the past, requiring manual regeneration of the ballot assets (which is an easy place to make a mistake).

Is it possible to instead generate these assets automatically (e.g. on demand when used, or on application startup)? E.g. a blank page image data could be generated in memory, a ballot image could be rendered from an election def.

File: libs/ui/src/ui_strings/app_strings.tsx (hunk 3 above)
Comment (jonahkagan) [reply libs/ui/src/ui_strings/app_strings.tsx:362]: We could remove the second sentence all together, it doesn't feel that necessary to me

File: libs/ui/src/diagnostics/mark_scan_readiness_report.test.tsx [hunk 13]
```diff
@@ -41,7 +48,9 @@ test('MarkScanReadinessReport', () => {
   screen.getByText(/All Precincts/);
   screen.getByText('Battery Level: 50%');
   screen.getByText('Power Source: Battery');
-  screen.getByText('Detected');
-  screen.getByText(/Test passed/);
+  screen.getByTestId('paperHandlerDetected');
+  screen.getByTestId('accessibleControllerDetected');
+  screen.getByTestId('paperHandlerTestPassed');
+  screen.getByTestId('accessibleControllerTestPassed');
```
Comment (jonahkagan) [inline libs/ui/src/diagnostics/mark_scan_readiness_report.test.tsx:53]: A pattern I've used in these cases is to find the section title, get its container using `.closest('section')` or similar, and then use `within` to find the text I'm looking for within that container. A bit more robust than using a test id since you still test the actual content displayed

File: libs/ui/src/diagnostics/mark_scan_paper_handler_section.tsx (hunk 7 above)
Comment (kshen0) [reply libs/ui/src/diagnostics/mark_scan_paper_handler_section.tsx:22]: https://github.com/votingworks/vxsuite/pull/4784

File: apps/mark-scan/frontend/src/pages/diagnostics/paper_handler_diagnostic_screen.tsx [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+      case 'paper_handler_diagnostic.prompt_for_paper':
+      case 'paper_handler_diagnostic.load_paper':
+        // Reuse appStrings value because we can, but other strings
+        // don't need to be appStrings because this page isn't voter-facing
+        contents = <P>{appStrings.instructionsBmdLoadPaper()}</P>;
+        break;
+      case 'paper_handler_diagnostic.print_ballot_fixture':
+      case 'paper_handler_diagnostic.scan_ballot':
+      case 'paper_handler_diagnostic.interpret_ballot':
+        contents = <P>A sample test ballot is being printed and scanned.</P>;
```
Comment (jonahkagan) [inline apps/mark-scan/frontend/src/pages/diagnostics/paper_handler_diagnostic_screen.tsx:44]: I think you can replace "sample test ballot" with "test ballot" to simplify

File: apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx [hunk 15]
```diff
@@ -108,6 +131,17 @@ export function DiagnosticsScreen({
...
         </Screen>
       </Route>
+      <Route path="/paper-handler">
+        <PaperHandlerDiagnosticScreen
+          onClose={() => {
+            history.push('/');
+            // The diagnostic record is written by the backend after successful rear ejection.
+            // Invalidating the query at the time of the last mutation in this flow is still too early
+            // so we have to manually refetch.
+            void mostRecentPaperHandlerDiagnosticQuery.refetch();
```
Comment (jonahkagan) [inline apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx:147]: Any reason to refetch here rather than invalidate?

Comment (kshen0) [reply apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx:147]: In testing, invalidation resulted in stale data being rendered. Only a hard refetch fixed it. I reviewed the state machine and I think this comment is outdated because the diagnostic record is written on entry to `success` or `failure` states.

I can play with this some more when I'm in front of the hardware next week but would prefer to leave this as-is since we are guaranteed it'll work.

Comment (jonahkagan) [reply apps/mark-scan/frontend/src/pages/diagnostics/diagnostics_screen.tsx:147]: Gotcha. I would recommend at least using `await` instead of `void`, since `void` swallows errors
